<?xml version='1.0'?>

<robot name="myrobot" xmlns:xacro="http://www.ros.org/wiki/xacro">



  <xacro:include filename="$(find slammer)/urdf/slammer.gazebo" />
  <xacro:include filename="$(find slammer)/urdf/materials.xacro" />
  <xacro:include filename="$(find slammer)/urdf/macros.xacro" />

  <xacro:property name="HPI" value="1.5707"/>

  <link name='chassis'>
    <pose>0 0 0 0 0 0</pose>

    <inertial>
      <mass value="1"/>
      <origin xyz="0.0 0 0.05" rpy=" 0 0 0"/>
      <box_inertia m="1" x="0.12" y="0.181" z ="0.05"/>
      <!--inertia
          ixx="0.5" ixy="0" ixz="0"
          iyy="1.0" iyz="0"
          izz="0.1"
      /-->

    </inertial>

    <collision name='collision'>
      <geometry>
        <box size=".12 .181 .05"/>
      </geometry>
    </collision>

    <visual name='chassis_visual'>
      <origin xyz="0 0 0" rpy=" 0 0 0"/>
      <geometry>
        <box size=".12 .181 .05"/>
      </geometry>
    </visual>



  </link>


  <link name="left_wheel">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.05" length="0.026"/>
      </geometry>
    </collision>
    <visual name="left_wheel_visual">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.05" length="0.026"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <mass value="0.03"/>
      <cylinder_inertia m="0.03" r="0.05" h="0.026"/>
      <!--inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/-->
    </inertial>
  </link>

  <link name="right_wheel">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.05" length="0.026"/>
      </geometry>
    </collision>
    <visual name="right_wheel_visual">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.05" length="0.026"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <mass value="0.03"/>
      <cylinder_inertia m="0.03" r="0.05" h="0.026"/>
      <!--inertia
        ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0"
        izz="1.0"/-->
    </inertial>
  </link>


  <joint type="continuous" name="left_wheel_joint">
    <origin xyz="0.04 0.1035 0" rpy="0 0 0"/>
    <child link="left_wheel"/>
    <parent link="chassis"/>
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="100" velocity="21"/>
    <joint_properties damping="1.0" friction="1.0"/>
  </joint>

  <joint type="continuous" name="right_wheel_joint">
    <origin xyz="0.04 -0.1035 0" rpy="0 0 0"/>
    <child link="right_wheel"/>
    <parent link="chassis"/>
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="100" velocity="21"/>
    <joint_properties damping="1.0" friction="1.0"/>
  </joint>









  <transmission name="transmission_left_wheel">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="left_wheel_joint">
          <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
      <actuator name="motor_left_wheel">
          <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
  </transmission>

  <transmission name="transmission_right_wheel">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="right_wheel_joint">
          <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
      <actuator name="motor_right_wheel">
          <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
          <mechanicalReduction>1</mechanicalReduction>
      </actuator>
  </transmission>











  <link name="rear_left_wheel">
        <collision name="collision">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.03" length="0.01"/>
      </geometry>
    </collision>
    <visual name="left_wheel_visual">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.03" length="0.01"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <mass value="0.03"/>
      <cylinder_inertia m="0.03" r="0.03" h="0.01"/>
    </inertial>
  </link>

  <link name="rear_right_wheel">
    <collision name="collision">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.03" length="0.01"/>
      </geometry>
    </collision>
    <visual name="right_wheel_visual">
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <geometry>
        <cylinder radius="0.03" length="0.01"/>
      </geometry>
    </visual>
    <inertial>
      <origin xyz="0 0 0" rpy="0 ${HPI} ${HPI}"/>
      <mass value="0.03"/>
      <cylinder_inertia m="0.03" r="0.03" h="0.01"/>

    </inertial>
  </link>





  <joint type="continuous" name="rear_left_wheel_joint">
    <origin xyz="-0.04 0.06 -0.02" rpy="0 0 0"/>
    <child link="rear_left_wheel"/>
    <parent link="chassis"/>
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="100" velocity="21"/>
    <joint_properties damping="1.0" friction="1.0"/>
  </joint>

  <joint type="continuous" name="rear_right_wheel_joint">
    <origin xyz="-0.04 -0.06 -0.02" rpy="0 0 0"/>
    <child link="rear_right_wheel"/>
    <parent link="chassis"/>
    <axis xyz="0 1 0" rpy="0 0 0"/>
    <limit effort="100" velocity="21"/>
    <joint_properties damping="1.0" friction="1.0"/>
  </joint>





  <joint name="laser_joint" type="fixed">
    <axis xyz="0.0 1 0.0" />
    <origin xyz="0.04 0 0.075" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="laser_link"/>
  </joint>


  <link name="laser_link">
    <collision>
      <origin xyz="0.0 0 -0.025" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.05"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0.0 0 -0.025" rpy="0 0 0"/>
      <geometry>
        <box size = "0.02 0.02 0.05"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="0.20" />
      <origin xyz="0.0 0 -0.025" rpy="0 0 0"/>
      <!--inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" /-->
      <box_inertia m="0.2" x="0.02" y="0.02" z="0.05"/>
    </inertial>
  </link>




  <joint name="imu_joint" type="fixed">
    <axis xyz="0 0 0"/>
    <origin xyz="0 0 0"/>
    <parent link="chassis"/>
    <child link="imu_link"/>
  </joint>


  <link name="imu_link">
    <inertial>
      <mass value="0.001"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <box_inertia m="0.001" x="0.001" y="0.001" z="0.001"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size=".001 .001 .001"/>
      </geometry>
    </collision>
  </link>



</robot>

